### PR TITLE
Refactor plugins in an ECS based way

### LINKED ISSUE #9: Remove Boost dependency
### Description

The whole rmf codebase has almost no boost dependency except for this package. We should try to remove any boost dependency since it's a bulky library and is only very sparingly used.

### Implementation considerations

Most of the functionality implemented in boost functions in this package is inside `thumbnail_generator` and `crowd_simulator` and has actually been migrated to the standard library, such as `filesystem`, `dynamic_pointer_cast` and exceptions.
The only non-immediate migration will be from `boost::program_options`, but there seem to be alternatives (such as the C style `getopt()`).

### LINKED ISSUE #95: Make lifts manage their own doors
### Description

Make lift simulation plugins manage their own doors, instead of using door plugins.

### Implementation considerations

Currently, lift doors are commanded through a `door` plugin. While this works, the architecture is fundamentally wrong since the simulated lifts will publish a request to the `door_supervisor` to request opening a lift door, which runs a door plugin.
In reality, lifts would manage their own doors and wouldn't need to go through RMF to operate them.
Additionally, since lift doors are running door plugins they also appear in the rviz door panel, which shouldn't be the case since RMF is not managing the doors individually, i.e. in the hotel world from rmf_demos:

image

### PR DESCRIPTION
**Pull request by luca-della-vedova**
### Implemented feature

Based on #113, fixes #95, fixes #9.
Needs https://github.com/open-rmf/rmf_traffic_editor/pull/483

### Implementation description

This PR ports the codebase to Harmonic, drops Gazebo Classic support and takes the chance to remove the common / gz split, which allows us to cleanup the codebase a fair bit and take advantage of ECS patterns.
Specifically, when using Gazebo classic, we were forced to adopt the same architecture of one plugin per object, leading to large inefficiencies, for example for each door we would have one state publisher + one request subscriber that would check each incoming request.

This PR refactors plugins to be in an ECS based way. Now doors / lifts have a single simple plugin called `register_component` that adds a component that contains all their properties.
There is now a single site wide door / lift manager system that iterates through all entities with door / lift components and performs whatever logic is needed.

For a more indepth overview of  the architecture, there is a FOSSAsia 2023 talk [here](https://www.youtube.com/watch?v=4D7kdpKszF8).

Furthermore, state updates moved from being continuous to being event based, to reduce traffic flow and improve scalability.
Specifically, we use a new feature in ROS 2 Iron, Matched Callbacks, now:

* Whenever a new subscriber is detected, we send the full state of the system, slightly throttled to avoid data floods (one publish per simulation update cycle).
* Whenever a change in a door state is detected, we send a message with the new state.
* When none of the above is true, no messages are sent, reducing load on downstream subscribers.
* State QoS is now reliable with a greater history depth since now we can't really afford to lose messages.

Both these changes should greatly with scaling RMF to larger sites, since they will cut down both computation time and network traffic.

Bonus tasks:

- [x] Make lift updates event based as well.
- [ ] Refactor dispenser to be ECS based and change to event based.
- [ ] Refactor ingestor to be ECS based and change to event based.
- [ ] Refactor slotcar to be ECS based.
- [ ] Remove `rmf_robot_sim_common` package and unify plugins.

### CODE REVIEW

**Comment by luca-della-vedova** [comment]
I'm moving this out of draft.

This change greatly improves performance and scalability, at least as far as doors and lifts are concerned.
I tested it on the hotel world which has 27 doors and 2 lifts and observed a change in maximum RTF from 1.5x to 2.6x.
For reference, fully disabling doors and lifts brings the RTF to about 3.3x, so the overhead of running a large number of doors and lifts is not too large.
The gain is not taking into account the improvements in downstream nodes that come from the event based state update publishing and should reduce callbacks by a few orders of magnitude (depending on utilization of facilities), although arguably these callbacks don't tend to be too resource intensive.

There is some room for improvement in bringing all the ideas put in the door plugin into other plugins, as well as optimizing slotcar, but this can be a good starting point.

**Comment by luca-della-vedova** [comment]
To help speed up development I removed the event based logic in https://github.com/open-rmf/rmf_simulation/pull/114/commits/045ed07168d646b6c3ae4d7bbbf8cad7c1aacf0c, this will have a performance impact but should keep the behavior of the rest of the stack unchanged.

File: rmf_building_sim_gz_plugins/src/register_component.cpp
Comment (mxgrey) [inline rmf_building_sim_gz_plugins/src/register_component.cpp:207]: Nothing to change here, just a remark: This approach of serializing components in the SDF is probably how SDF should be getting used in general. It would be worth discussing upstream whether the SDF maintainers would consider first-class support for this pattern.

File: rmf_robot_sim_gz_plugins/src/slotcar.cpp
Comment (mxgrey) [inline rmf_robot_sim_gz_plugins/src/slotcar.cpp:195]: This is a much better approach to filtering these entities 👍 👍

**Review by mxgrey — CHANGES_REQUESTED** [review]
I've tested this against 24.04 and rolling, and everything seems to be working.

I left some in-line comments for places where changes are needed, just to prevent segfaults and/or confusing crashes that could conceivably happen in some edge cases. If I find an opportunity, I'll make the changes myself and push them.

**Review by mxgrey — APPROVED** [review]
I've made the recommended changes. Now it would be good if we can get CI working before merging, but if that's blocked by upstream issues that won't be resolved soon then I suppose we can move forward with merging and sort out the CI when things are ready upstream.

File: rmf_building_sim_gz_plugins/src/lift.cpp
Comment (luca-della-vedova) [reply rmf_building_sim_gz_plugins/src/lift.cpp]: The logic in the corner case here survived a long time without ever being caught since it's in the case where the initial floor is not specified, which doesn't happen in any of our pipelines.
This was supposed to be `max`, and below we would either iterate through all the floors and find the one with the minimum elevation, _or_ get the first element in the list. However, this snippet did both, and in the wrong way (`lift.floors` is an `unordered_map` with `std::string` as a key so I think `lift.floors.at(0)` would actually look for the lift named `0` and fail).
I addressed all of these in https://github.com/open-rmf/rmf_simulation/pull/114/commits/54b559ce4d8946f4ba488f1a229f35cf7383e383, where now the variable is initialized to `max` just to make sure the for loop will reliably find the minimum, then initialize the value at `0` if no floor is found as you pointed out.
I'm not 100% convinced with initializing `target_elevation` with 0 since doing a `if (floor.elevation < target_elevation)` would only trigger if the floor elevation was below 0, which might not be the case if we are only simulating upper floors.

File: rmf_building_sim_gz_plugins/src/door.cpp
Comment (luca-della-vedova) [reply rmf_building_sim_gz_plugins/src/door.cpp:200]: Yap nice find!

File: rmf_building_sim_gz_plugins/src/lift.cpp
Comment (luca-della-vedova) [reply rmf_building_sim_gz_plugins/src/lift.cpp]: Applied the proper fix denoted in https://github.com/open-rmf/rmf_simulation/pull/114#discussion_r1611748364 and got rid of the `at` in the first place, which was actually wrong

**Review by mxgrey — APPROVED** [review]

**Comment by luca-della-vedova** [comment]
Jazzy CI failure is expected until all rmf repos have a Jazzy brach

**Pull request merged**
